### ISSUE TITLE
Regex filter to block random ads/trackers

### ISSUE THREAD

**Issue by ghajini** [easylist/easylist#6476]
### List the website(s) you're having issues:
URL | 3rd party to block
--- | ---
`https://1123movies.to/episodes/watch-the-mandalorian-season-2-episode-1-online/` | `https://ahaumplab.com/rsHtYfC82ow6YN/14280`
`https://1123movies.to/episodes/watch-the-mandalorian-season-2-episode-1-online/` | `https://rilelogicbuy.com/fcU9CXGXtAGYOoA/17354`
`https://9xupload.xyz/f/51uogp-0h2j2z61svrf9` | `https://frogwokshive.com/rIvIK4QHh6h0nJUh/24142`
`https://streamwire.net/vu5vupr41sso` | `https://umekian.pw/rAs6OdIJC7sjWBcD6/16255`
`https://skymovieshd.nl`: | `https://dostophog.com/r8ezT5VyQHbuMI6/12336`
`https://clicknupload.co/gn1xjxnokso3` | `https://vimjarawl.com/f3MwYQ6fPan/22621`
`https://uptomega.me/49rf3ow09i12` | `https://ledmophemp.com/rnkHzsMoBCNP5k/12790`
`https://megaup.net/2fFbi/RWDSKMHD_(2020)_www.SkymoviesHD.nl_720p_HDRip_Hindi_Dubbed_x264_AAC.mkv` | `https://mehebborc.com/tZSXTlgKmZrdJXTSv/18410`
`https://uploader.link/f/o0dmub-y89jy75oisxk8` | `https://scripsjeermint.com/rlvKAUqId4GiV9Z/15626`
`https://dailyuploads.net/fkzecjp9yqed` | `https://cogleapad.com/ftnSjcx7ErWNg4XS/16949`
`https://abcvideo.cc/embed-e4mmrcla64ov.html` | `https://eveteaemo.com/rdr0tIIVodZhg95hY/23763`
`https://www.123movie.show` | `https://baleiambwee.com/fSCu8UmlMMOKDU/27641`
`https://123movies.mx/movies/roald-dahls-the-witches/` | `https://tithewtog.com/riRW2mGqr0VkZO/15636`

**& so on there may be many sites such 3rd party ads may exist .../r... or .../f.... on movie sites and file sharing websites**

### What happens?
3rd party random ads/trackers
make regexp rule to block these ads as they seem to be from same ad company i guess

### List Subscriptions you're using:

ublock origin default lists

### Your settings

- OS/version: win10
- Browser/version: firefox, opera
- Adblock Extension/version: 1.30.6

### Other details:
i do aware of there are already rules in easylist which block these ads but my concern is about unknown ads/trackers which may not been blocked by easylist as of now

i do aware of such regex rule in easylist `/^https?:\/\/[0-9a-z]{11,20}.*.com\/.*/$3p,domain=9xmovies.app|9xupload.xyz|mixdrop.co|mixdrop.to|ovagames.com`

may be you can modify this rule & make generalised efficiently so that each time easylist don't have to block tracker individually....

@ryanbr 
@mapx- ,@okiehsch @Yuki2718

**Comment by Yuki2718** [comment]
How about `/^https://[a-z]{7,14}\.com/[0-9a-zA-Z]{14,17}/\d{5}$/$script,3p`?

**Comment by ghajini** [comment]
@Yuki2718 
It doesn't block

`https://vimjarawl.com/f3MwYQ6fPan/22621` 

On `https://clicknupload.co/gn1xjxnokso3`

**Comment by Yuki2718** [comment]
Adjusted: `/^https://[a-z]{7,15}\.(?:com|pw)/[0-9a-zA-Z]{11,17}/\d{4,5}$/$script,3p`
Will investigate later on.

**Comment by Yuki2718** [comment]
Another case here: https://github.com/AdguardTeam/AdguardFilters/issues/68043
`https://puttrotgob.com/fwGJIhEoZnlfkjPC/26481`

**Comment by ghajini** [comment]
Filter looks good as of now but since its regex filter, let's wait

**Comment by Yuki2718** [comment]
I wonder if the filter can be added to uAssets if it takes too long for EL to adopt it. Is it possible @mapx- @uBlock-user @okiehsch ?  #6484 could proactively be blocked if the regex was there and I think `(TLDs)/[0-9a-zA-Z]{11,17}/\d{4,5}$` is relatively strong condition for JS to prevent FP.

**Comment by ghajini** [comment]
@Yuki2718 browse some free download movies websites in Japan to test filter( search on yahoo for movie sites )....I finished my search almost all sites, no FP reported yet

**Comment by Yuki2718** [comment]
@ghajini Done shortly, I too see no FP so far. Thx mapx- for adding the filter!

**Comment by ghajini** [comment]
proven effective
at
`https://nbyts.online/15-times-people-asked-korean-photoshop-masters-for-help-and-regretted-it`
on `https://vaneplusnod.com/t904ZMn6GHDwDam/24359`

**Comment by gwarser** [comment]
Making sure first part of path includes at least one of lower case, upper case and digit instead of just any from this set may help, but this will be some higher level regex stuff.

**Comment by keczuppp** [comment]
> Yuki2718 : I too see no FP so far

- it was just a matter of time (several days) until such broad regex will begin to generate false positives and break legitimate websites as happened so there: https://github.com/uBlockOrigin/uAssets/issues/8280
- but here I was surprised, I found another similiar risky regex in Fanboy Enhanced Tracking list:
`/:\/\/www\.([a-z]{6,16})\.com\/([a-z]{1,4})\.js$/$script,third-party`
example theoretical / synthetic false hit: `https://www.thirdparty.com/test.js`
but I've been always using `FET` list, long time already and never experienced a single breakage caused by the risky regex, interesting

**Comment by Yuki2718** [comment]
@keczuppp I'd say actually scripts ending with `/[0-9a-zA-Z]{11,17}/\d{4,5}$` is not so common, 99% of legitimate scripts end with `.js`. Do you think `/^https?:\/\/[a-z]{7,15}\.(?:com|pw)\/(?=[0-9A-Z]{0,16}[a-z])(?=[0-9a-z]{0,16}[A-Z])(?=[a-zA-Z]{0,16}[0-9])[0-9a-zA-Z]{11,17}\/\d{4,5}$/$script,match-case,3p`; i.e. having at least one capital letter and number each, still cause FP and have an example?

**Comment by gwarser** [comment]
Path will still match words starting with upper case and ending with a number - not sure this is really uncommon :/

**Comment by Yuki2718** [comment]
At least two capital then? And the last digits are not totally random, so:
`/^https?:\/\/[a-z]{7,15}\.(?:com|pw)\/(?=[0-9A-Z]{0,16}[a-z])(?=[0-9a-z]{0,15}[A-Z][0-9a-z]{0,15}[A-Z])[0-9a-zA-Z]{11,17}\/(?:[12]\d{4}|[5-9]\d{3})$/$script,3p`

**Comment by Yuki2718** [comment]
The first path is also not totally random, always starts either from f, r, or t. This filter
`/^https?:\/\/[a-z]{7,15}\.(?:com|pw)\/[frt](?=[0-9A-Z]{0,15}[a-z])(?=[0-9a-z]{0,14}[A-Z][0-9a-z]{0,14}[A-Z])[0-9a-zA-Z]{10,16}\/(?:[12]\d{4}|[5-9]\d{3})$/$script,3p,match-case`
will significantly reduce the chance of FP while catche all these known ad-scripts.

**Comment by keczuppp** [comment]
- yep, I have been just about to write a comment about it but I see you found it by yourself already...
- however there is still one more thing: need to add `site` domain ==> `(?:com|pw|site)`
see ghajini: https://github.com/easylist/easylist/issues/6476#issuecomment-730518380 : `http://ifsmatbeg.site/rQ20h7yTXB2QSX/12755`

**Comment by keczuppp** [comment]
So lets summarise, the random pattern has the following constants:
- starts with `[frt]`
- ends with `(?:[12]\d{4}|[5-9]\d{3})$`
- has at least 2 capital letters
- I also noticed that the pattern has at least 3 lower case letters, 
does you regex include it already?, if not can it be added?
- just a notice: the pattern sometimes contains no numbers, so we never should consider numbers as a constant

**Comment by Yuki2718** [comment]
>   does you regex include it already?, if not can it be added?

No, because it's not very useful to distinguish from legit requests (most of requests matching to other parts of the regex will include 3+ lower case in the path anyway). As regex is heavy, better to focus only on useful characteristics.

> * just a notice: the pattern sometimes contains no digits, so we never should consider digits as a constant

If the above filter still caused the FP, allowing some false negative may not be avoidable.

**Comment by keczuppp** [comment]
Yes you definately have a krystian3w syndrome, replying so fast at the time I have been still editing my comment.

> keczuppp :  does you regex include it already?, if not can it be added?
>> Yuki2718 : No, because it's not very useful to distinguish from legit requests (most of requests matching to other parts of the regex will include 3+ lower case in the path anyway). As regex is heavy, better to focus only on useful characteristics.

Sure, you've replied at the same time I just have realised it myself and have been editing / correcting my comment: 
keczuppp : https://github.com/easylist/easylist/issues/6476#issuecomment-737185713 : "although not sure whether it is so important"

> keczuppp :  just a notice: the pattern sometimes contains no digits, so we never should consider digits as a constant
>> Yuki2718 : If the above filter still caused FP, allowing some false negative may not be avoidable.

Sure, you have replied at the time I have been editing "the pattern" to "the middle pattern", which mean I'm not talking about digits at the end, hence I don't know to which digits you are reffering with your reply.

**Comment by Yuki2718** [comment]
Ha-ha, I see you meant digits in the part starting from frt. ;)

**Comment by keczuppp** [comment]
- okay, so we have all the important constants implemented into the regex at the moment
- the last thing I'm gonna do now is to collect more examples of the 3rd party request to block to see whether the pattern will change or not

**Comment by Yuki2718** [comment]
I'm appreciated for that. You know, though, we have to wait for match-case option to come to stable release of uBO.

**Comment by keczuppp** [comment]
- I haven't found many of new requests, just few, no change, so whoever is interested 
(the people who watch videos on such websites) can try to find more and see if anything changes
- or just to leave it as is, if anyone will find anything accidentaly, the regex can be customized at any time if necessary
- anyway the regex seems to be completed & ready at the moment

> Yuki2718 : I'm appreciated for that. You know, though, we have to wait for match-case option to come to stable release of uBO.

- true in case of adding the regex into `uBO` filters
- but we are currently on `EasyList` github tracker...
so it is not true in case of adding the regex into `EasyList`, 
as `ABP` already does support `match-case`

**Comment by Yuki2718** [comment]
I heard ABP dropped support, IDK. Anyway, it's going to off-topic. I remember to add site TLD.

**Comment by keczuppp** [comment]
>  Yuki2718  : I heard ABP dropped support, IDK.

- then you have bad info..because `match-case` is just working fine in `ABP` (v3.10)
unless you meant it is dropped in some dev release

**Comment by ghajini** [comment]
few sites where iam using this regex `/^https?:\/\/[a-z]{7,15}\.(?:com|pw)\/(?=[0-9A-Z]{0,16}[a-z])(?=[0-9a-z]{0,16}[A-Z])[0-9a-zA-Z]{11,17}\/\d{4,5}$/$script,3p,match-case` 
are 
```
https://www.mr-jatt.im/punjabi-music/song/494908/Keh-Len-De-Kaka-mp3-song.html?nhcnt
https://9xupload.asia/f/brfcaw-vgwbjdwnlges7
http://uppom.live/g2fsd9oug3o8
https://www43.zippyshare.com/v/jSHkIwQI/file.html
https://driveapp.in/file/h69yqegJ1T
https://howblogs.xyz/ee48b4
https://ww1.put-locker.com
```
some sites are using random ads ,this rule saves my time until ryanbr adds

**Comment by mapx-** [comment]
We should still wait to be sure `match-case` arrived for nearly all installations, otherwise there are FP

**Comment by ghajini** [comment]
yes iam using in dev build ,also lot of test required b4 adding this rule,cheers for last regex rule 👍

**Comment by keczuppp** [comment]
> mapx- & ghajini: https://github.com/easylist/easylist/issues/6476#issuecomment-740012329We should still wait to be sure `match-case` arrived for nearly all installations, otherwise there are FP

- not true, if a regex rule with `match-case` is used in older uBO versions which don't support it, the whole rule will be just rejected, therefore no false hits

**Comment by Yuki2718** [comment]
Yes, the final version will be `/^https?:\/\/[a-z]{7,15}\.(?:com|pw|site)\/[frt](?=[0-9A-Z]{0,15}[a-z])(?=[0-9a-z]{0,14}[A-Z][0-9a-z]{0,14}[A-Z])[0-9a-zA-Z]{10,16}\/(?:[12]\d{4}|[5-9]\d{3})$/$script,3p,match-case`
We probably can't make it more strict, as there's already an "only 2 capital" case: `https://truernewlywait.com/r5aKtrmWa5nd/16895`. Having tested with links provided by @ghajini , it catches all. OTOH, I also confirmed these pages constantly changes domains e.g. `site.mkvmoviesking.net` is now `film.mkvmoviesking.com`. This means even with denyallow it's hard to catch them up, rationalizing the use of regex.

**Comment by keczuppp** [comment]
> Yuki2718 : `3p`

- should be `third-party` to be compatible and added into `EasyList` , like I already mentioned: https://github.com/easylist/easylist/issues/6476#issuecomment-737251588
- also we are on `EasyList` github tracker, like I already mentioned: https://github.com/easylist/easylist/issues/6476#issuecomment-737251588

> Yuki2718 : We probably can't make it more strict, as there's already an "only 2 capital" case: `https://truernewlywait.com/r5aKtrmWa5nd/16895.`

- true, but If mentioning one reason, should mention the second as well:

> keczuppp : https://github.com/easylist/easylist/issues/6476#issuecomment-737185713 :
> just a notice: the middle pattern sometimes contains no digits, so we never should consider digits as a constant
> ( example : `https://emolapnay.com/rlnzLzvjGRKBY/20600`, `https://mehebborc.com/tZSXTlgKmZrdJXTSv/18410` )

<br>

> Yuki2718 : Having tested with links provided by @ghajini , it catches all. 

- agree, I tested it 2 days ago already, but since there was no change and the regex still catches all of them, I didn't even bother to mention about it unless anything changes...

>  Yuki2718 : This means even with denyallow it's hard to catch them up, rationalizing the use of regex.

- agree, also the regex is completed, we are awaiting a decision whether it will be added or not, into `EasyList` ,
if it will be declined to be added into `EasyList` we will be awaiting a decision to be or not, added into `uBlock filters`

**Comment by Yuki2718** [comment]
@keczuppp EL authors and us uAssets contributors are closely communicating to one another not only on GH but also in Slack. In fact you see some issues fixed on uAssets were closed on EL repo for the fix and vice versa. EL authors are aware of all these discussion so if they think it's appropriate to add the filter then they will, but I have no idea, maybe they won't do anytime soon for the risk of FP. OTOH @mapx- and I can add the filter to uAssets at anytime, just waiting for match-case to be available for all uBO installation. Still any info about false positive/negative are welcome, it's best to improve the filter before adding it.

**Comment by keczuppp** [comment]
> Yuki2718 : @keczuppp EL authors and us uAssets contributors are closely communicating to one another not only on GH but also in Slack. In fact you see some issues fixed on uAssets were closed on EL repo for the fix and vice versa. EL authors are aware of all these discussion so if they think it's appropriate to add the filter then they will, but I have no idea, maybe they won't do anytime soon for the risk of FP

Great, as if I didn't know..., but it is good to let them know the regex has been completed, also by "decision" it means we can wait for the commit, a week for example, and not necessarily until they write back in the comment...

**Comment by Yuki2718** [comment]
@keczuppp I didn't assume you're noob. Sorry if my saying made you feel like that.

**Comment by keczuppp** [comment]
ok bro

**Comment by Yuki2718** [comment]
I'm finding this family of scripts are more widely used then I thought. Domains added in https://github.com/uBlockOrigin/uAssets/commit/58e96626bd23efd40d85764dcd2bfc7356960302 are covered. `selfswayjay.com` at `www.animefenix.com` too. But I found a case the last digits are `4823` at `https://www.liivideo.com/embed-ykcz07t13hse.html` so now `/^https?:\/\/[a-z]{7,15}\.(?:com|pw|site)\/[frt](?=[0-9A-Z]{0,15}[a-z])(?=[0-9a-z]{0,14}[A-Z][0-9a-z]{0,14}[A-Z])[0-9a-zA-Z]{10,16}\/(?:[12]\d{4}|[4-9]\d{3})$/$script,3p,match-case`

**Comment by ghajini** [comment]
@Yuki2718 
On this site `https://dvdplay.run/` Easylist is addressing the above ads but blocking one ad generates another ad.....

So now with Easylist disabled and your rule in place
![Screenshot_2020-12-25-00-45-36-985_org mozilla fenix](https://user-images.githubusercontent.com/20338483/103104037-fa8cfc80-464a-11eb-8df0-104e4eba0434.png)

Ads are missed, can it be readjusted or new rule???

**Comment by Yuki2718** [comment]
@ghajini So domain can include numbers, and the maximum length also have to be extended. Now it's `/^https?:\/\/[0-9a-z]{7,16}\.(?:com|pw|site)\/[frt](?=[0-9A-Z]{0,15}[a-z])(?=[0-9a-z]{0,14}[A-Z][0-9a-z]{0,14}[A-Z])[0-9a-zA-Z]{10,16}\/(?:[12]\d{4}|[4-9]\d{3})$/$script,3p,match-case`.

**Comment by ghajini** [comment]
with `/^https?:\/\/[0-9a-z]{7,16}\.(?:com|pw|site)\/[frt](?=[0-9A-Z]{0,15}[a-z])(?=[0-9a-z]{0,14}[A-Z][0-9a-z]{0,14}[A-Z])[0-9a-zA-Z]{10,16}\/(?:[12]\d{4}|[4-9]\d{3})$/$script,3p,match-case`
in place it didn't catch `https://f3551539d6565853.com/r6NnmWbQYUo/14703` at `dvdplay.run`

**Comment by Yuki2718** [comment]
You're right, however, a question is how prevalent xhr is - so far this and https://github.com/easylist/easylist/pull/6791. I think this regex is not tokenizable and adding xhr support means more requests to check and potentially more chance of false positive.

**Comment by ghajini** [comment]
May be leave xhr ad crap task to easylist only
also
how to abort script which is at top in page which is generating these craps using scriptlets?...

**Comment by Yuki2718** [comment]
I believe you meant "how about"?

**Comment by mapx-** [comment]
@ghajini 

try
`dvdplay.run##+js(aost, Math, onreadystatechange, 2)`

**Comment by Yuki2718** [comment]
@mapx- How about `rawmanga.top##+js(acis, addEventListener, XMLHttpRequest)` for rawmanga? Anyway I leave these to you.

**Comment by ghajini** [comment]
got like this @mapx- 
![image](https://user-images.githubusercontent.com/20338483/103137323-fcf46280-46ed-11eb-9099-27d050819da2.png)

**Comment by mapx-** [comment]
@Yuki2718 working fine, it's the same inline script as for dvdplay but is not at the top

**Comment by mapx-** [comment]
@ghajini yes, it's maximum we can do

or this 1:
`dvdplay.run##+js(aopw, XMLHttpRequest)`

but it's safer aost one to not breaking something else

**Comment by Yuki2718** [comment]
> didn't get what you mean to say?

I thought you probably knew *how to* block but suggesting us to add scriptlet filters.

**Comment by ghajini** [comment]
yeah i have in place beforehand `dvdplay.run##+js(aopr, XMLHttpRequest)` but it was same situation like https://github.com/easylist/easylist/issues/6476#issuecomment-751258716 ,so i asked

**Comment by Yuki2718** [comment]
Ah, I got. It's fortunate for me that mapx- is here, as I couldn't answer either ;)

**Comment by mapx-** [comment]
I added the regex above slightly modified (removed `[frt]`), tested here
`https://www.watchseries1.video/`

**Comment by Yuki2718** [comment]
@mapx- Hmm...I personally think [frt] is a good measure to prevent possible FP, as these scripts always start from one of them. It may also reduce steps for mismatch i.e. non-matching request will be dropped before proceeding to expensive lookahead part. Maybe you care about future change by them?

**Comment by mapx-** [comment]
> I now understand script on your link start from p

yeah, exactly

**Comment by ghajini** [comment]
Issue fixed with using ublock origin ,closing

**Comment by Yuki2718** [comment]
I noticed the aost filter is not yet added. I'll add it.

**Comment by ghajini** [comment]
Just a side note here
`https://1todaypk.ws/12-o-clock-2021-hindi-full-movie-watch-online-free-download-14456`

clicking play generates popup request of above crap
![image](https://user-images.githubusercontent.com/20338483/104206118-2d116480-5455-11eb-915b-9e783d980b32.png)

 which is not yet filtered by regex

**Comment by Yuki2718** [comment]
@ghajini We may add $popup only if it becomes common and can't catch them up well.

**Comment by ghajini** [comment]
yes for testing now as a custom filter

**Comment by ghajini** [comment]
`https://ww18.watchmovierulz.me/special-ops-2020-telugu-tamil-hindi-malayalam-kan/full-movie-watch-online-free-1-925.html`

popups
blocking `popup` request in regex will help

or

**Comment by stephenhawk8054** [comment]
@Yuki2718 There's another script

`https://mesalplayboy.com/fS20wHghQFyuO0s0/60350`

in https://github.com/AdguardTeam/AdguardFilters/issues/140055

Is it good to change `[1-5]` -> `[1-6]`

```adb
/^https?:\/\/(?:[a-z]{2}\.)?[0-9a-z]{7,16}\.com\/[a-z](?=[a-z]{0,25}[0-9A-Z])[0-9a-zA-Z]{3,26}\/(?:[1-6]\d{4}|[3-9]\d{3})\??(?:_=\d+|v=\d)?$/$frame,script,xhr,popup,3p,match-case
```

?

**Comment by Yuki2718** [comment]
> Is it good to change

Sure, why not?

**Comment by stephenhawk8054** [comment]
> > Is it good to change
> 
> Sure, why not?

Yeah, I was just not sure if there were FPs in the past that needed to be `[1-5]`
